PublicGitArchive/pga/pga/pga.go

```diff
@@ -3,202 +3,70 @@
-func (r *Repository) ToCSV() []string {
-	return []string{
-		headerURL:               r.URL,
-		headerSivaFilenames:     formatStringList(r.Filenames),
-		headerFileCount:         formatInt(r.Files),
-		headerLangs:             formatStringList(r.Languages),
-		headerLangsByteCount:    formatIntList(r.LanguagesByteCount),
-		headerLangsLinesCount:   formatIntList(r.LanguagesLineCount),
-		headerLangsFilesCount:   formatIntList(r.LanguagesFileCount),
-		headerCommitsCount:      formatInt(r.Commits),
-		headerBranchesCount:     formatInt(r.Branches),
-		headerForkCount:         formatInt(r.Forks),
-		headerEmptyLinesCount:   formatIntList(r.LanguagesEmptyLines),
-		headerCodeLinesCount:    formatIntList(r.LanguagesCodeLines),
-		headerCommentLinesCount: formatIntList(r.LanguagesCommentLines),
-		headerLicense:           r.License,
-		headerStars:             formatInt(r.Stars),
-		headerSize:              formatInt(r.Size),
+func (e *errorBadHeaderLength) Error() string {
+	if e.expectedMin == e.expectedMax {
+		return fmt.Sprintf("bad header length: expected  %d but got %d", e.expectedMin, e.length)
 	}
+	return fmt.Sprintf("bad header length: expected  %d to %d but got %d", e.expectedMin, e.expectedMax, e.length)
 }
 
-// Index provides iteration over repositories.
-type Index interface {
-	Next() (*Repository, error)
-}
-
-type csvIndex struct {
-	r         *csv.Reader
-	withStars bool
-	withSize  bool
+type errorBadHeaderColumn struct {
+	expected string
+	index    int
+	col      string
 }
 
-func (idx *csvIndex) Next() (*Repository, error) {
-	rows, err := idx.r.Read()
-	if err != nil {
-		return nil, err
-	}
-	if rows[0] != csvHeaders[0] {
-		if !idx.withStars {
-			rows = append(rows, "-1")
-		}
-
-		if !idx.withSize {
-			rows = append(rows, "-1")
-		}
-
-		return RepositoryFromCSV(rows)
-	}
-	return idx.Next()
+func (e *errorBadHeaderColumn) Error() string {
+	return fmt.Sprintf("bad header column: expected  %s at index %d but got %s", e.expected, e.index, e.col)
 }
 
-var errBadHeader = fmt.Errorf("bad header, expected %s", strings.Join(csvHeaders, ","))
-
-// IndexFromCSV returns an Index reading from a CSV file.
-func IndexFromCSV(r io.Reader) (Index, error) {
-	cr := csv.NewReader(r)
-	headers, err := cr.Read()
-	if err != nil {
-		return nil, fmt.Errorf("could not skip headers row: %v", err)
-	}
-
-	// check for compatibility between old indexes
-	length := len(headers)
-	expected := len(csvHeaders)
-	if length < expected-2 || length > expected {
-		return nil, errBadHeader
-	}
-
-	var withStars, withSize bool
-	for i := range headers {
-		h := headers[i]
-		if h != csvHeaders[i] {
-			return nil, errBadHeader
-		}
-
-		switch h {
-		case csvHeaders[headerStars]:
-			withStars = true
-		case csvHeaders[headerSize]:
-			withSize = true
-		}
-	}
-
-	return &csvIndex{cr, withStars, withSize}, nil
+// Index provides iteration over repositories.
+type Index interface {
+	Next() (Repository, error)
```

Comment:
I am totally confused at this point. What is going on with Index, Repository and Dataset, what is their relationship?

It is more common to provide `ForEach(func(r Repository))` instead of `Next()`.

Reply:
We should have crystal clear entities. We should also rip away "Repository" from everywhere because it confuses very much (Git? nope, gotcha).

Depending on how you describe the current logic, I can suggest the appropriate abstractions.